Automattic/wp-calypso
# Checkout: Pass contact info to updateCartContactDetails directly

#### Changes proposed in this Pull Request

When you click to complete the contact info/billing info step of checkout, that data is saved to the shopping cart using the function `updateCartContactDetails()` (for tax purposes). That function is memoized using `useCallback()` because it needs to remain stable for use as a dependency. The function saves the data it finds within its closure. Unfortunately, this closure makes it somewhat difficult to reason about the state of that data as it updates.

In this PR we modify `updateCartContactDetails()` (renamed to `updateCartContactDetailsForCheckout()`) so that it now gets the contact info data directly as an argument. This way we do not depend on the value in that function's memoized closure. We actually pass in all of its required data instead of relying on a closure for any of it.

In addition, we used to actively erase the contact details in the cart when the payment method changed to `free-purchase` (see https://github.com/Automattic/wp-calypso/pull/42905) but I realized that this has a flaw: the `free-purchase` payment method only appears when the cart total is 0, in which case no taxes would be charged anyway. Therefore this diff removes that behavior (it made more sense when we also ignored taxes for full-credits purchases but [we no longer do that](https://github.com/Automattic/wp-calypso/pull/47641)).

#### Testing instructions

- Add a product to your cart and visit checkout.
- Open the network tab in your browser's devtools and monitor requests to the `/me/shopping-cart` endpoint.
- Edit your contact details/billing info to fill it in (or change it). Press "Continue" on that step.
- Verify that a `/me/shopping-cart` network request is made and that it includes the country code/postal code you entered.
